Cookbook/Proposals/SmarterDiffs.md

```diff
@@ -0,0 +1,97 @@
+```
+
+As can be seen from the diff it does not show in what exact array item the failure happened. This is something that we will need to adjust along with [this issue](https://github.com/krzysztofzablocki/Difference/issues/1). 
+
+Proposed way to integrate this feature is to copy the implementation (with proper author attribution) and adjust it to our needs, optionally we can make pull requests to the original library.
+
+## Impact on existing codebase
+
+This approach can be rolled out gradually and used for new test (Xcode templates will use a new assert) or tests can be refactored in batches by replacing existing assertions with a new assert.
+
+## Alternatives considered
+
+As an alternative we can enforce asserting on individual states instead of all states which will make diffs less of an issue. But as mentioned before it will not eliminate it.
+
+Another alternative would be to use snapshot testing that we already use for UI tests. It will make our assert just a call to a single `assertSnapshot` function. But at the same time it will mean that all the snapshots will be written into the files rather than explicitly stated in the test code. This will decrease the boilerplate we will have to write in tests manually but will make it harder to understand test expectations as as we will need to search through text files to find values we need.
```

Comment:
I wonder if it would be possible to use snapshot library infrastructure to capture diffs on the fly in memory without storing them in files. It would work like `Difference` as in produce snapshots only on failure.
Would require some more usage to see how useful are the diffs it provides. I think we have less experience with this than with `Difference`, so it's hard to tell.

Reply:
The convenience of snapshots is that we don't have to specify the expected values since they are in written the file. If we don't write them then we have to construct them on every run so we end up with the same as regular asserts, maybe a bit different API.